### ISSUE TITLE
x/tools/gopls: feature Request: fill struct 

### ISSUE THREAD

**Issue by frioux** [golang/go#37576]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.14 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

n/a (feature request)

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/frew/.cache/go-build"
GOENV="/home/frew/.config/go/env"
GOEXE=""
GOFLAGS="-mod=readonly"
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/frew/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org"
GOROOT="/home/frew/sdk/go1.14"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/frew/sdk/go1.14/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/dev/null"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/run/shm/go-build445574914=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What did you do?

n/a (feature request)

### What did you expect to see?

n/a (feature request)

### What did you see instead?

n/a (feature request)

### Feature request

vim-go had a command called `:GoFillStruct` and it would transform (for example), this:

```go
x := Person{}
```

into

```go
x := Person{
	Name: "",
	Employed: false,
	Age: 0,
}
```

Basically it'd set all the fields to their zero value.  In https://github.com/govim/govim/issues/709 I was told that this needs to be added to gopls first, so here I am :)

**Comment by stamblerre** [comment]
@muirdm: Do you think we could fit this behavior into completion?

**Comment by muirdm** [comment]
For my understanding, why is this useful behavior?

**Comment by stamblerre** [comment]
I don't know if it's particularly useful for the `Person` case described above, but I would find it useful if the struct contained map values. (My classic example is `go/types.Info`, which I always find super annoying to create.) My ideal scenario when completing that struct is that, if I have a value of the correct type in scope, it should be used -- otherwise, we should make a map in the struct declaration.

**Comment by muirdm** [comment]
The zero value of map is "nil", so this proposal as written wouldn't help in the types.Info case.

In my experience having the literal completion that gives you a candidate for `make(map[some.Long]annoying.Map)` is already a major improvement. Having structs get automatically filled in with in-scope objects sounds hard to avoid false positives.

**Comment by frioux** [comment]
the main use case here is creating fresh struct with lots of fields. This
is common when working with, for example, the aws sdk where completing
methods is a very small help since you then need to pass in a struct with
sometimes dozens of fields set.

**Comment by muirdm** [comment]
I can see how that could be part of your workflow, but it also seems to overlap with existing facilities such as textDocument/hover which will show you all the struct fields or just a normal completion request in the struct literal, which also will list all the struct fields and their types.

To answer the original question of fitting this into completion, we could but I'm not sure what would trigger the completion or how it would show up clearly to the user. Maybe better as a code action?

**Comment by frioux** [comment]
I don't know what a code action is, but in the original setting it was something the user would explicitly request, rather than fitting into the same workflow as method completion.  I personally think that's fine; this is something I'd do a few times a day, not constantly, like method completion.

**Comment by frioux** [comment]
I don't know what a code action is, but in the original setting it was something the user would explicitly request, rather than fitting into the same workflow as method completion.  I personally think that's fine; this is something I'd do a few times a day, not constantly, like method completion.

**Comment by leitzler** [comment]
I support the `codeAction` suggestion, seems like it would be a better fit (maybe of kind  `refactor.rewrite.fillstruct`).

As a feature this is something that, at least I, hear people ask for quite often. With the use case to get the field names quickly added when creating a new struct instance. Hover does show the same info, but it would be really helpful if gopls could add it for you so that you don't have to add them manually.

**Comment by luciolas** [comment]
Is anyone working on this? Mind if I take a look at implementing this feature?

**Comment by stamblerre** [comment]
I don't believe so, so please go ahead, @luciolas! As a starting point, I would suggest implementing this behavior as a [command](https://github.com/golang/tools/blob/44c82bac18acc91bc88d9ec6c29c434a2617b0e9/internal/lsp/command.go#L18).

**Comment by luciolas** [comment]
Ahh, I went ahead and wrote some code in code_action.go. Is there a reason for implementing it as a command? 

Secondly, I also tested if completions are viable for a struct fill. 
(See GIF below for a small demo on using code action and completions)

![X1Lgc2noZR](https://user-images.githubusercontent.com/22808115/78646719-0df50d80-78ec-11ea-96ab-9d997adf5665.gif)

Thirdly, a few things I have in mind:
1. Does Code Action support placeholders? 
2. Is it possible to trigger completion while in a placeholder?

**Comment by stamblerre** [comment]
> Ahh, I went ahead and wrote some code in code_action.go. Is there a reason for implementing it as a command?

I think either way probably works. I wasn't sure if it necessarily fit under `"refactor.rewrite"`, but I think it works.

> Secondly, I also tested if completions are viable for a struct fill.
> (See GIF below for a small demo on using code action and completions)

Looks great! Thanks for working on this. I think the code action is a good starting point, and we could possible extend this to be part of completions if we think of a good user experience for it.

> Thirdly, a few things I have in mind:
> 
> 1. Does Code Action support placeholders?

I don't believe so.

> 2. Is it possible to trigger completion while in a placeholder?

I think it should be possible, yes.

**Comment by a-h** [comment]
This looks great, I'd also like to see a code action to fill a struct from a variable (merge).

```go
type A struct {
  ID string
  Name string
  Phone string
  Email string
}

type B struct {
  Name string
  Phone string
  Email string
  Fax string
}
```

In this case, the snippet would need to know the source and destination variable names (they should be in scope). Given variables `a` and `b` of type `A` and `B` respectively, the snippet would produce something like this.

```
var a A
// Not present on B.
// a.ID = b.ID
a.Name = b.Email
a.Phone = b.Email
a.Email = b.Email
// Not present on A.
// a.Fax = b.Fax
```

This would prevent a common mistake of forgetting to copy a field from one type to another, missing a field, or overwriting the wrong names (e.g. a.Fax = b.Email). These types of things pop up as the "last line effect", where developers switch off as they keep track of the fields they have or haven't mapped.

Runtime tools exist such as https://github.com/imdario/mergo but these must use reflection, resulting in poorer performance, and the use of a 3rd party library.

Not sure if this should be a completely different thing (`merge`?), but this feature looks like a good starting point.
